File: src/investraton/ingest/etf_holdings.py
"""ETF composition ingestion for look-through exposure.

Primary free source: iShares (BlackRock) daily holdings CSV — predictable URLs,
full constituent list with weight + sector + country. Compositions are cached as
JSON under data/etf_holdings/ so we don't refetch constantly and degrade
gracefully offline. Non-iShares ETFs can be supplied via a manual JSON file of
the same shape.
"""

from __future__ import annotations

import csv
import io
import json
import logging
from datetime import datetime, timezone
from pathlib import Path

import requests
import yaml

from ..config import DATA_DIR, RESOURCE_CONFIG_DIR
from ..models import Constituent, ETFComposition

log = logging.getLogger(__name__)

CACHE_DIR = DATA_DIR / "etf_holdings"
_HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}


def base_ticker(ticker: str) -> str:
    """Strip an exchange suffix: IWDA.AS -> IWDA, CSPX.L -> CSPX."""
    return ticker.upper().split(".", 1)[0]


def _load_registry() -> dict:
    path = RESOURCE_CONFIG_DIR / "etf_sources.yaml"
    if not path.exists():
        return {}
    return yaml.safe_load(path.read_text(encoding="utf-8")) or {}


def _resolve_ishares(base: str, isin: str, registry: dict) -> dict | None:
    """Resolve a ticker/ISIN to a fetchable product: a direct iShares fund, or a
    non-iShares fund mapped to a same-index iShares fund (index proxy).

    Returns {"product_id", "name", "isin"} or None.
    """
    isin = (isin or "").upper().strip()
    for entry in registry.get("ishares", []) or []:
        aliases = {str(a).upper() for a in entry.get("aliases", [])}
        aliases.add(str(entry.get("product_id", "")).upper())
        if base in aliases or (isin and isin == str(entry.get("isin", "")).upper()):
            return {"product_id": entry["product_id"], "name": entry.get("name", base),
                    "isin": entry.get("isin", "")}
    for entry in registry.get("index_proxies", []) or []:
        aliases = {str(a).upper() for a in entry.get("aliases", [])}
        if base in aliases or (isin and isin == str(entry.get("isin", "")).upper()):
            return {"product_id": entry["proxy_product_id"], "name": entry.get("name", base),
                    "isin": entry.get("isin", "")}
    return None


def get_composition(ticker: str, isin: str = "") -> ETFComposition | None:
    """Return the composition for an ETF, or None if we have no source.

    Resolves by ticker alias or ISIN -> iShares product id -> holdings CSV.
    Order: fresh cache -> iShares fetch -> stale cache -> manual file.
    """
    base = base_ticker(ticker)
    registry = _load_registry()
    settings = registry.get("settings", {}) or {}
    max_age = settings.get("max_age_days", 7)
    entry = _resolve_ishares(base, isin, registry)
    cache_key = entry["product_id"] if entry else base

    cached = _read_cache(cache_key)
    if cached is not None and _cache_age_days(cache_key) <= max_age:
        return cached

    if entry:
        url = settings.get("url_template", "").format(pid=entry["product_id"])
        fetched = _fetch_ishares(cache_key, url)
        if fetched is not None:
            fetched.name = entry.get("name", base)
            _write_cache(fetched)
            return fetched

    if cached is not None:
        cached.stale = True
        return cached  # better stale than nothing

    manual = _read_manual(base)
    if manual is not None:
        return manual

    log.info("No ETF composition source for %s/%s (not in registry, no cache/manual).", base, isin)
    return None


# ---- iShares CSV ----
def _fetch_ishares(base: str, url: str) -> ETFComposition | None:
    try:
        r = requests.get(url, headers=_HEADERS, timeout=30)
        r.raise_for_status()
        text = r.content.decode("utf-8-sig")
    except Exception as e:  # noqa: BLE001 - external
        log.warning("iShares fetch failed for %s: %s", base, e)
        return None
    return _parse_ishares_csv(base, text)


def _parse_ishares_csv(base: str, text: str) -> ETFComposition | None:
    lines = text.splitlines()
    as_of = ""
    if lines and lines[0].lower().startswith("fund holdings as of"):
        parts = lines[0].split(",", 1)
        as_of = parts[1].strip().strip('"') if len(parts) > 1 else ""
    hdr_idx = next((i for i, l in enumerate(lines) if l.replace('"', "").startswith("Ticker,")), None)
    if hdr_idx is None:
        log.warning("iShares CSV for %s had no recognizable header.", base)
        return None

    reader = csv.DictReader(io.StringIO("\n".join(lines[hdr_idx:])))
    constituents: list[Constituent] = []
    for row in reader:
        tkr = (row.get("Ticker") or "").strip()
        weight = _pct(row.get("Weight (%)"))
        if not tkr or weight <= 0:
            continue
        constituents.append(
            Constituent(
                ticker=tkr,
                name=(row.get("Name") or "").strip(),
                weight=weight / 100.0,
                sector=(row.get("Sector") or "").strip(),
                country=(row.get("Location") or "").strip(),
                asset_class=(row.get("Asset Class") or "Equity").strip(),
            )
        )
    if not constituents:
        return None
    return ETFComposition(
        ticker=base, name=base, as_of=as_of or "unknown", source="ishares", constituents=constituents
    )


def _pct(v) -> float:
    try:
        return float(str(v).replace(",", "").replace("%", "").strip())
    except (TypeError, ValueError):
        return 0.0


# ---- cache + manual ----
def _cache_path(base: str) -> Path:
    return CACHE_DIR / f"{base}.json"


def _read_cache(base: str) -> ETFComposition | None:
    return _read_json(_cache_path(base), default_source="cache")


def _read_manual(base: str) -> ETFComposition | None:
    return _read_json(_cache_path(base), default_source="manual")


def _read_json(path: Path, default_source: str) -> ETFComposition | None:
    if not path.exists():
        return None
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
        return ETFComposition(
            ticker=raw["ticker"],
            name=raw.get("name", raw["ticker"]),
            as_of=raw.get("as_of", "unknown"),
            source=raw.get("source", default_source),
            constituents=[
                Constituent(
                    ticker=c["ticker"], name=c.get("name", c["ticker"]), weight=float(c["weight"]),
                    sector=c.get("sector", ""), country=c.get("country", ""),
                    asset_class=c.get("asset_class", "Equity"),
                )
                for c in raw.get("constituents", [])
            ],
        )
    except Exception as e:  # noqa: BLE001
        log.warning("Could not read composition file %s: %s", path, e)
        return None


def _write_cache(comp: ETFComposition) -> None:
    CACHE_DIR.mkdir(parents=True, exist_ok=True)
    payload = {
        "ticker": comp.ticker,
        "name": comp.name,
        "as_of": comp.as_of,
        "source": comp.source,
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "constituents": [
            {"ticker": c.ticker, "name": c.name, "weight": c.weight,
             "sector": c.sector, "country": c.country, "asset_class": c.asset_class}
            for c in comp.constituents
        ],
    }
    _cache_path(comp.ticker).write_text(json.dumps(payload), encoding="utf-8")


def _cache_age_days(base: str) -> float:
    path = _cache_path(base)
    if not path.exists():
        return 1e9
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
        fetched = datetime.fromisoformat(raw["fetched_at"])
        return (datetime.now(timezone.utc) - fetched).total_seconds() / 86400.0
    except Exception:  # noqa: BLE001
        return 1e9

File: engine/src/ingest/etfHoldings.ts
/** ETF composition ingestion — port of ingest/etf_holdings.py.
 *
 * Primary free source: iShares (BlackRock) daily holdings CSV — predictable
 * product-id URLs, full constituent list with weight + sector + country.
 * Non-iShares funds tracking the same index are resolved through
 * `index_proxies` in etf_sources.yaml (this is how EXUS, an Xtrackers fund,
 * gets look-through from an iShares fund on the same index).
 *
 * Compositions are cached as JSON so we don't refetch constantly and degrade
 * gracefully offline. Order: fresh cache -> fetch -> stale cache -> manual file.
 */

import { load as yamlLoad } from "js-yaml";

import type { Constituent, ETFComposition } from "../models.ts";
import type { FileSystem, HttpClient, Paths } from "../platform.ts";
import { dataDir, resourceConfigDir } from "../platform.ts";
import { parseCsvDicts } from "../store/files.ts";

export interface RegistryEntry {
  product_id?: string | number;
  proxy_product_id?: string | number;
  isin?: string;
  name?: string;
  aliases?: (string | number)[];
}

export interface Registry {
  ishares?: RegistryEntry[];
  index_proxies?: RegistryEntry[];
  settings?: { max_age_days?: number; url_template?: string };
}

export interface ResolvedProduct { product_id: string; name: string; isin: string }

/** Strip an exchange suffix: IWDA.AS -> IWDA, CSPX.L -> CSPX. */
export const baseTicker = (ticker: string): string => ticker.toUpperCase().split(".", 1)[0];

const cacheDir = (p: Paths, fs: FileSystem) => fs.join(dataDir(p, fs), "etf_holdings");
const cachePath = (p: Paths, fs: FileSystem, base: string) =>
  fs.join(cacheDir(p, fs), `${base}.json`);

export async function loadRegistry(p: Paths, fs: FileSystem): Promise<Registry> {
  const path = fs.join(resourceConfigDir(p, fs), "etf_sources.yaml");
  const text = await fs.readText(path);
  if (text === null) return {};
  return (yamlLoad(text) as Registry) ?? {};
}

/** Resolve a ticker/ISIN to a fetchable iShares product, directly or via an
 * index proxy. Mirrors _resolve_ishares(). */
export function resolveIshares(
  base: string, isin: string, registry: Registry,
): ResolvedProduct | null {
  const want = (isin || "").toUpperCase().trim();
  const aliasesOf = (e: RegistryEntry, extra?: unknown) => {
    const s = new Set((e.aliases ?? []).map((a) => String(a).toUpperCase()));
    if (extra !== undefined) s.add(String(extra ?? "").toUpperCase());
    return s;
  };
  for (const e of registry.ishares ?? []) {
    const aliases = aliasesOf(e, e.product_id);
    if (aliases.has(base) || (want && want === String(e.isin ?? "").toUpperCase())) {
      return { product_id: String(e.product_id), name: e.name ?? base, isin: e.isin ?? "" };
    }
  }
  for (const e of registry.index_proxies ?? []) {
    const aliases = aliasesOf(e);
    if (aliases.has(base) || (want && want === String(e.isin ?? "").toUpperCase())) {
      return { product_id: String(e.proxy_product_id), name: e.name ?? base, isin: e.isin ?? "" };
    }
  }
  return null;
}

/** float(str(v).replace(",","").replace("%","").strip()) with 0.0 on failure. */
export function pct(v: unknown): number {
  if (v === null || v === undefined) return 0;
  const n = Number(String(v).replace(/,/g, "").replace(/%/g, "").trim());
  return Number.isFinite(n) ? n : 0;
}

/** Parse an iShares daily holdings CSV. Mirrors _parse_ishares_csv(). */
export function parseIsharesCsv(base: string, text: string): ETFComposition | null {
  const lines = text.split(/\r\n|\r|\n/);
  let asOf = "";
  if (lines.length && lines[0].toLowerCase().startsWith("fund holdings as of")) {
    const idx = lines[0].indexOf(",");
    if (idx !== -1) asOf = lines[0].slice(idx + 1).trim().replace(/^"|"$/g, "");
  }
  const hdrIdx = lines.findIndex((l) => l.replace(/"/g, "").startsWith("Ticker,"));
  if (hdrIdx === -1) return null; // no recognizable header

  const rows = parseCsvDicts(lines.slice(hdrIdx).join("\n"));
  const constituents: Constituent[] = [];
  for (const row of rows) {
    const tkr = (row["Ticker"] ?? "").trim();
    const weight = pct(row["Weight (%)"]);
    if (!tkr || weight <= 0) continue;
    constituents.push({
      ticker: tkr,
      name: (row["Name"] ?? "").trim(),
      weight: weight / 100,
      sector: (row["Sector"] ?? "").trim(),
      country: (row["Location"] ?? "").trim(),
      asset_class: (row["Asset Class"] ?? "Equity").trim() || "Equity",
    });
  }
  if (constituents.length === 0) return null;
  return {
    ticker: base, name: base, as_of: asOf || "unknown",
    source: "ishares", constituents, stale: false,
  };
}

// ---- cache + manual ----
interface CachedPayload {
  ticker: string; name?: string; as_of?: string; source?: string;
  fetched_at?: string;
  constituents?: Array<Partial<Constituent> & { ticker: string; weight: number | string }>;
}

async function readJsonComposition(
  p: Paths, fs: FileSystem, base: string, defaultSource: string,
): Promise<{ comp: ETFComposition; fetchedAt: string | null } | null> {
  const text = await fs.readText(cachePath(p, fs, base));
  if (text === null) return null;
  try {
    const raw = JSON.parse(text) as CachedPayload;
    const comp: ETFComposition = {
      ticker: raw.ticker,
      name: raw.name ?? raw.ticker,
      as_of: raw.as_of ?? "unknown",
      source: raw.source ?? defaultSource,
      constituents: (raw.constituents ?? []).map((c) => ({
        ticker: c.ticker,
        name: c.name ?? c.ticker,
        weight: Number(c.weight),
        sector: c.sector ?? "",
        country: c.country ?? "",
        asset_class: c.asset_class ?? "Equity",
      })),
      stale: false,
    };
    return { comp, fetchedAt: raw.fetched_at ?? null };
  } catch {
    return null; // unreadable composition file
  }
}

/** Age of a cached composition in days; 1e9 when missing/unreadable. */
export function cacheAgeDays(fetchedAt: string | null, now = Date.now()): number {
  if (!fetchedAt) return 1e9;
  const t = Date.parse(fetchedAt);
  if (!Number.isFinite(t)) return 1e9;
  return (now - t) / 86_400_000;
}

async function writeCache(
  p: Paths, fs: FileSystem, comp: ETFComposition,
): Promise<void> {
  await fs.mkdirp(cacheDir(p, fs));
  const payload = {
    ticker: comp.ticker, name: comp.name, as_of: comp.as_of, source: comp.source,
    fetched_at: new Date().toISOString(),
    constituents: comp.constituents.map((c) => ({
      ticker: c.ticker, name: c.name, weight: c.weight,
      sector: c.sector, country: c.country, asset_class: c.asset_class,
    })),
  };
  await fs.writeText(cachePath(p, fs, comp.ticker), JSON.stringify(payload));
}

/** Return the composition for an ETF, or null if we have no source at all. */
export async function getComposition(
  ticker: string,
  isin: string,
  ctx: { paths: Paths; fs: FileSystem; http: HttpClient },
): Promise<ETFComposition | null> {
  const { paths, fs, http } = ctx;
  const base = baseTicker(ticker);
  const registry = await loadRegistry(paths, fs);
  const settings = registry.settings ?? {};
  const maxAge = settings.max_age_days ?? 7;
  const entry = resolveIshares(base, isin, registry);
  const cacheKey = entry ? entry.product_id : base;

  const cached = await readJsonComposition(paths, fs, cacheKey, "cache");
  if (cached && cacheAgeDays(cached.fetchedAt) <= maxAge) return cached.comp;

  if (entry) {
    const url = (settings.url_template ?? "").replace("{pid}", entry.product_id);
    const text = url ? await http.getText(url) : null;
    const fetched = text ? parseIsharesCsv(cacheKey, text) : null;
    if (fetched) {
      fetched.name = entry.name ?? base;
      await writeCache(paths, fs, fetched);
      return fetched;
    }
  }

  if (cached) return { ...cached.comp, stale: true }; // better stale than nothing

  const manual = await readJsonComposition(paths, fs, base, "manual");
  return manual ? manual.comp : null;
}
